src/Microsoft.Health.Dicom.SqlServer/Registration/Experimental/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,177 @@
+using Microsoft.Health.SqlServer.Configs;
+using Microsoft.Health.SqlServer.Extensions;
+using Microsoft.Health.SqlServer.Features.Client;
+using Microsoft.Health.SqlServer.Features.Storage;
+using Polly;
+using Polly.Contrib.WaitAndRetry;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class ServiceCollectionExtensions
+    {
+        private const string SqlServerSectionName = "SqlServer";
+
+        public static IServiceCollection AddSqlServer(
+            this IServiceCollection services,
+            IConfiguration configurationRoot,
+            Action<ISqlServiceBuilder> configureSqlServer = null)
+        {
+            EnsureArg.IsNotNull(services, nameof(services));
+            EnsureArg.IsNotNull(configurationRoot, nameof(configurationRoot));
+
+            var config = new SqlServerDataStoreConfiguration();
+            configurationRoot.GetSection(SqlServerSectionName).Bind(config);
+
+            services
+                .AddSingleton(config)
+                .AddConnectionServices(config);
+
+            configureSqlServer?.Invoke(new SqlServiceBuilder(services));
+
+            return services;
+        }
+
+        private static IServiceCollection AddConnectionServices(
+            this IServiceCollection services,
+            SqlServerDataStoreConfiguration config)
+        {
+            services.Add<SqlTransactionHandler>()
+                .Scoped()
+                .AsSelf()
+                .AsImplementedInterfaces();
+
+            services.Add<SqlServerTransientFaultRetryPolicyFactory>()
+                .Singleton()
+                .AsSelf()
+                .AsImplementedInterfaces();
+
+            services.Add<RetrySqlCommandWrapperFactory>()
+                .Singleton()
+                .AsSelf()
+                .AsService<SqlCommandWrapperFactory>();
+
+            services.Add<SqlConnectionWrapperFactory>()
+                .Scoped()
+                .AsSelf()
+                .AsImplementedInterfaces();
+
+            services.AddSingleton<ISqlConnectionStringProvider, DefaultSqlConnectionStringProvider>();
+
+            switch (config.AuthenticationType)
+            {
+                case SqlServerAuthenticationType.ManagedIdentity:
+                    services.AddSingleton<ISqlConnectionFactory, ManagedIdentitySqlConnectionFactory>();
+                    services.AddSingleton<IAccessTokenHandler, ManagedIdentityAccessTokenHandler>();
+                    services.AddSingleton<AzureServiceTokenProvider>();
+                    break;
+                case SqlServerAuthenticationType.ConnectionString:
+                default:
+                    services.AddSingleton<ISqlConnectionFactory, DefaultSqlConnectionFactory>();
+                    break;
+            }
+
+            return services;
+        }
+
+        #region Shared Component Copies (TODO: Remove)
+
+        private sealed class RetrySqlCommandWrapper : SqlCommandWrapper
```

Comment:
do we want to move these class into separate partial file so that this file is more clean?

Reply:
This is all temporary and is only used in this class for set-up. I'll keep them here for now, and of course delete them before merging into main